README.md
# Tatter\Settings
Lightweight settings management for CodeIgniter 4

[![](https://github.com/tattersoftware/codeigniter4-settings/workflows/PHPUnit/badge.svg)](https://github.com/tattersoftware/codeigniter4-settings/actions/workflows/test.yml)
[![](https://github.com/tattersoftware/codeigniter4-settings/workflows/PHPStan/badge.svg)](https://github.com/tattersoftware/codeigniter4-settings/actions/workflows/analyze.yml)
[![](https://github.com/tattersoftware/codeigniter4-settings/workflows/Deptrac/badge.svg)](https://github.com/tattersoftware/codeigniter4-settings/actions/workflows/inspect.yml)
[![Coverage Status](https://coveralls.io/repos/github/tattersoftware/codeigniter4-settings/badge.svg?branch=develop)](https://coveralls.io/github/tattersoftware/codeigniter4-settings?branch=develop)

## Quick Start

1. Install with Composer: `> composer require tatter/settings`
2. Update the database: `> php spark migrate -all`
3. Use `spark` to create templates: `> php spark settings:add timezone user America/New_York`
4. Use the service to access user settings:
```
service('settings')->timezone = $_POST['timezone_preference'];
...
$userTimezone = service('settings')->timezone;
```

## Features

Provides ready-to-use settings management for CodeIgniter 4

## Installation

Install easily via Composer to take advantage of CodeIgniter 4's autoloading capabilities
and always be up-to-date:
* `> composer require tatter/settings`

Or, install manually by downloading the source files and adding the directory to
`app/Config/Autoload.php`.

Once the files are downloaded and included in the autoload, run any library migrations
to ensure the database is setup correctly:
* `> php spark migrate -all`

## Usage

Once the library is included all the resources are ready to go and you are ready to start
adding settings. You may import setting templates directly into the `settings` table or
add them manually with the CLI command `php spark settings:add`.

**Settings** also comes with a database seeder for some recommended default templates. Run
the seeder from the command line:
	
	php spark db:seed "Tatter\Settings\Database\Seeds\SettingsSeeder"

This will add appropriately-scoped templates and default values for the following settings:

| Name             | Description                                     | Data Type | Default Value                   | Protected |
| ---------------- | ----------------------------------------------- | --------- | ------------------------------- | --------- |
| siteVersion      | Current version of this project                 | string    | 1.0.0                           | Yes       |
| brandName        | Brand name for this project                     | string    | Brand                           | Yes       |
| brandLogo        | Brand logo for this project                     | string    | /assets/images/logo.png         | Yes       |
| orgName          | Your organization name                          | string    | Organization                    | Yes       |
| orgLogo          | Your organization logo                          | string    | /assets/images/logo.png         | Yes       |
| orgUrl           | Your organization URL                           | uri       | https://example.com             | Yes       |
| orgAddress       | Your organization address                       | string    | 4141 Postmark Dr  Anchorage, AK | Yes       |
| orgPhone         | Your organization phone                         | string    | (951) 262-3062                  | Yes       |
| currencyUnit     | Currency format for number helper               | string    | USD                             | Yes       |
| currencyScale    | Conversion rate to the fractional monetary unit | int       | 100                             | Yes       |
| databaseTimezone | Timezone for the database server(s              | string    | UTC                             | Yes       |
| serverTimezone   | Timezone for the web server(s)                  | string    | UTC                             | Yes       |
| timezone         | Timezone for the user                           | string    | America/New_York                | No        |
| theme            | Site display theme                              | int       | 1                               | No        |
| perPage          | Number of items to show per page                | int       | 10                              | No        |

*Warning: This list is subject to change between major versions.*

Note that the seeder will not overwrite existing values so it is safe to re-run at any time.
See also [src/Database/Seeds/SettingsSeeder.php](src/Database/Seeds/SettingsSeeder.php).

### Setting Scope

``Settings`` come in three modes: global, user, and dynamic.
* Global settings are the same for every user and provide project owners to set application-wide values; set `protected` to `1`
* User settings start with a template value but each user may make their own value that persists across sessions; set `protected` to `0`
* Dynamic settings have no template but can be created and returned on-the-fly; they only persists for the current session.

Examples:

| Name          | Scope   | Content          | Notes                                        | Protected? |
|-------------- | ------- | ---------------- | -------------------------------------------- | ---------- |
| latestRelease | Global  | 0.7.6            | Git-style tag of latest code release         | 1          |
| timezone      | User    | America/New_York | Local timezone to use across the application | 0          |
| perPage       | User    | 10               | Default number of items to show per page     | 0          |
| jobsSearch    | Dynamic | backend php      | User's most recent search term for jobs      | n/a        |


* When you release a new version of your software:

	model('SettingModel')->where('name', 'latestRelease')->update(['content' => '1.2.3');
	echo view('home', ['latestRelease' => service('settings')->latestRelease]);

* When a user searches a list of jobs:

	$settings->jobsSearch = $_POST['searchTerm'];
	$data['jobs'] = $jobModel->paginate($settings->perPage);

### Magic Config

``Settings`` comes with a magic configuration file that allows direct access to template values. This is a convenient
way to access the library in a traditional framework fashion:

	$logo = config('Settings')->projectLogo;

Note that unlike the Service or Library values from the magic config are directly from the template default and are not
affected by user overrides:

	service('settings')->set('perPage', 20);

	echo service('settings')->perPage; // 20
	echo config('Settings')->perPage; // 10


CHANGELOG.md
# Changelog

## [v2.0.1](https://github.com/tattersoftware/codeigniter4-settings/tree/v2.0.1) (2021-07-10)

[Full Changelog](https://github.com/tattersoftware/codeigniter4-settings/compare/v2.0.0...v2.0.1)

**Closed issues:**

- Authentication Optional [\#24](https://github.com/tattersoftware/codeigniter4-settings/issues/24)
- Writing values to database [\#20](https://github.com/tattersoftware/codeigniter4-settings/issues/20)

**Merged pull requests:**

- Prep for Release [\#28](https://github.com/tattersoftware/codeigniter4-settings/pull/28) ([MGatner](https://github.com/MGatner))
- Coding Style [\#27](https://github.com/tattersoftware/codeigniter4-settings/pull/27) ([MGatner](https://github.com/MGatner))
- Refresh toolkit [\#26](https://github.com/tattersoftware/codeigniter4-settings/pull/26) ([MGatner](https://github.com/MGatner))
- Remove auth requirement [\#25](https://github.com/tattersoftware/codeigniter4-settings/pull/25) ([MGatner](https://github.com/MGatner))
- Demonstrate failing array casts. [\#23](https://github.com/tattersoftware/codeigniter4-settings/pull/23) ([sfadschm](https://github.com/sfadschm))
- Remove Users [\#19](https://github.com/tattersoftware/codeigniter4-settings/pull/19) ([MGatner](https://github.com/MGatner))

## [v2.0.0](https://github.com/tattersoftware/codeigniter4-settings/tree/v2.0.0) (2021-05-26)

[Full Changelog](https://github.com/tattersoftware/codeigniter4-settings/compare/v1.2.1...v2.0.0)

**Closed issues:**

- User Session ID [\#13](https://github.com/tattersoftware/codeigniter4-settings/issues/13)
- Cache key [\#9](https://github.com/tattersoftware/codeigniter4-settings/issues/9)

**Merged pull requests:**

- Prep for release [\#18](https://github.com/tattersoftware/codeigniter4-settings/pull/18) ([MGatner](https://github.com/MGatner))
- Magic Config [\#17](https://github.com/tattersoftware/codeigniter4-settings/pull/17) ([MGatner](https://github.com/MGatner))
- Refactor [\#16](https://github.com/tattersoftware/codeigniter4-settings/pull/16) ([MGatner](https://github.com/MGatner))

## [v1.2.1](https://github.com/tattersoftware/codeigniter4-settings/tree/v1.2.1) (2021-05-24)

[Full Changelog](https://github.com/tattersoftware/codeigniter4-settings/compare/v1.2.0...v1.2.1)

**Merged pull requests:**

- Refresh Components [\#15](https://github.com/tattersoftware/codeigniter4-settings/pull/15) ([MGatner](https://github.com/MGatner))
- Update Toolkit [\#14](https://github.com/tattersoftware/codeigniter4-settings/pull/14) ([MGatner](https://github.com/MGatner))

## [v1.2.0](https://github.com/tattersoftware/codeigniter4-settings/tree/v1.2.0) (2021-03-07)

[Full Changelog](https://github.com/tattersoftware/codeigniter4-settings/compare/v1.1.0...v1.2.0)

**Closed issues:**

- Trying to get property of non-object  [\#7](https://github.com/tattersoftware/codeigniter4-settings/issues/7)
- Bug : function getUser in settings.php [\#5](https://github.com/tattersoftware/codeigniter4-settings/issues/5)
- Bug: Cache system not working [\#3](https://github.com/tattersoftware/codeigniter4-settings/issues/3)

**Merged pull requests:**

- Bump toolkit [\#8](https://github.com/tattersoftware/codeigniter4-settings/pull/8) ([MGatner](https://github.com/MGatner))

## [v1.1.0](https://github.com/tattersoftware/codeigniter4-settings/tree/v1.1.0) (2020-11-12)

[Full Changelog](https://github.com/tattersoftware/codeigniter4-settings/compare/v1.0.5...v1.1.0)

**Closed issues:**

- Bug: Cache separator [\#2](https://github.com/tattersoftware/codeigniter4-settings/issues/2)
- Bug: Cache [\#1](https://github.com/tattersoftware/codeigniter4-settings/issues/1)

**Merged pull requests:**

- Development Tools, Dependencies [\#4](https://github.com/tattersoftware/codeigniter4-settings/pull/4) ([MGatner](https://github.com/MGatner))

## [v1.0.5](https://github.com/tattersoftware/codeigniter4-settings/tree/v1.0.5) (2019-09-26)

[Full Changelog](https://github.com/tattersoftware/codeigniter4-settings/compare/v1.0.4...v1.0.5)

## [v1.0.4](https://github.com/tattersoftware/codeigniter4-settings/tree/v1.0.4) (2019-07-03)

[Full Changelog](https://github.com/tattersoftware/codeigniter4-settings/compare/v1.0.3...v1.0.4)

## [v1.0.3](https://github.com/tattersoftware/codeigniter4-settings/tree/v1.0.3) (2019-06-20)

[Full Changelog](https://github.com/tattersoftware/codeigniter4-settings/compare/v1.0.2...v1.0.3)

## [v1.0.2](https://github.com/tattersoftware/codeigniter4-settings/tree/v1.0.2) (2019-06-12)

[Full Changelog](https://github.com/tattersoftware/codeigniter4-settings/compare/v1.0.1...v1.0.2)

## [v1.0.1](https://github.com/tattersoftware/codeigniter4-settings/tree/v1.0.1) (2019-04-09)

[Full Changelog](https://github.com/tattersoftware/codeigniter4-settings/compare/v1.0.0...v1.0.1)

## [v1.0.0](https://github.com/tattersoftware/codeigniter4-settings/tree/v1.0.0) (2019-04-08)

[Full Changelog](https://github.com/tattersoftware/codeigniter4-settings/compare/4c48b42d14507d61cff0456b8550e446dbaf92c2...v1.0.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


src/Database/Seeds/SettingSeeder.php
<?php

/**
 * This file is part of Tatter Settings.
 *
 * (c) 2021 Tatter Software
 *
 * For the full copyright and license information, please view
 * the LICENSE file that was distributed with this source code.
 */

namespace Tatter\Settings\Database\Seeds;

use CodeIgniter\Database\Seeder;
use RuntimeException;
use Tatter\Settings\Models\SettingModel;

class SettingSeeder extends Seeder
{
    public function run()
    {
        // Define default project setting templates
        $rows = [
            [
                'name'      => 'siteVersion',
                'datatype'  => 'string',
                'summary'   => 'Current version of this project',
                'content'   => '1.0.0',
                'protected' => 1,
            ],
            [
                'name'      => 'brandName',
                'datatype'  => 'string',
                'summary'   => 'Brand name for this project',
                'content'   => 'Brand',
                'protected' => 1,
            ],
            [
                'name'      => 'brandLogo',
                'datatype'  => 'string',
                'summary'   => 'Brand logo for this project',
                'content'   => '/assets/images/logo.png',
                'protected' => 1,
            ],
            [
                'name'      => 'orgName',
                'datatype'  => 'string',
                'summary'   => 'Your organization name',
                'content'   => 'Organization',
                'protected' => 1,
            ],
            [
                'name'      => 'orgLogo',
                'datatype'  => 'string',
                'summary'   => 'Your organization logo',
                'content'   => '/assets/images/logo.png',
                'protected' => 1,
            ],
            [
                'name'      => 'orgUrl',
                'datatype'  => 'uri',
                'summary'   => 'Your organization URL',
                'content'   => 'https://example.com',
                'protected' => 1,
            ],
            [
                'name'      => 'orgAddress',
                'datatype'  => 'string',
                'summary'   => 'Your organization address',
                'content'   => '4141 Postmark Dr  Anchorage, AK',
                'protected' => 1,
            ],
            [
                'name'      => 'orgPhone',
                'datatype'  => 'string',
                'summary'   => 'Your organization phone',
                'content'   => '(951) 262-3062',
                'protected' => 1,
            ],
            [
                'name'      => 'currencyUnit',
                'datatype'  => 'string',
                'summary'   => 'Currency format for number helper',
                'content'   => 'USD',
                'protected' => 1,
            ],
            [
                'name'      => 'currencyScale',
                'datatype'  => 'int',
                'summary'   => 'Conversion rate to the fractional monetary unit',
                'content'   => '100',
                'protected' => 1,
            ],
            [
                'name'      => 'databaseTimezone',
                'datatype'  => 'string',
                'summary'   => 'Timezone for the database server(s)',
                'content'   => 'UTC',
                'protected' => 1,
            ],
            [
                'name'      => 'serverTimezone',
                'datatype'  => 'string',
                'summary'   => 'Timezone for the web server(s)',
                'content'   => 'UTC',
                'protected' => 1,
            ],
            [
                'name'      => 'timezone',
                'datatype'  => 'string',
                'summary'   => 'Timezone for the user',
                'content'   => 'America/New_York',
                'protected' => 0,
            ],
            [
                'name'      => 'theme',
                'datatype'  => 'int',
                'summary'   => 'Site display theme',
                'content'   => '1',
                'protected' => 0,
            ],
            [
                'name'      => 'perPage',
                'datatype'  => 'int',
                'summary'   => 'Number of items to show per page',
                'content'   => '10',
                'protected' => 0,
            ],
        ];

        // Check for and create project setting templates
        foreach ($rows as $row) {
            if (! $setting = model(SettingModel::class)->where('name', $row['name'])->first()) {
                // No match - add the row
                if (! model(SettingModel::class)->allowCallbacks(false)->insert($row)) {
                    throw new RuntimeException(implode('. ', model(SettingModel::class)->errors())); // @codeCoverageIgnore
                }
            }
        }
    }
}
